suave/backends/redis_pubsub.go

```diff
@@ -0,0 +1,177 @@
+package backends
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/flashbots/go-utils/cli"
+	"github.com/go-redis/redis/v8"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+	suave "github.com/ethereum/go-ethereum/suave/core"
+)
+
+var (
+	redisUpsertTopic = "store:upsert"
+
+	redisConnectionPoolSize = cli.GetEnvInt("REDIS_CONNECTION_POOL_SIZE", 0) // 0 means use default (10 per CPU)
+	redisMinIdleConnections = cli.GetEnvInt("REDIS_MIN_IDLE_CONNECTIONS", 0) // 0 means use default
+	redisReadTimeoutSec     = cli.GetEnvInt("REDIS_READ_TIMEOUT_SEC", 0)     // 0 means use default (3 sec)
+	redisPoolTimeoutSec     = cli.GetEnvInt("REDIS_POOL_TIMEOUT_SEC", 0)     // 0 means use default (ReadTimeout + 1 sec)
+	redisWriteTimeoutSec    = cli.GetEnvInt("REDIS_WRITE_TIMEOUT_SEC", 0)    // 0 means use default (3 seconds)
+
+	formatRedisBidKey = func(bidId suave.BidId) string {
+		return fmt.Sprintf("bid-%x", bidId)
+	}
+
+	formatRedisBidValueKey = func(bidId suave.BidId, key string) string {
+		return fmt.Sprintf("bid-data-%x-%s", bidId, key) // TODO: should also include the hash of the bid at least
+	}
+)
+
+type RedisPubSub struct {
+	ctx      context.Context
+	cancel   context.CancelFunc
+	redisUri string
+	client   *redis.Client
+	pubsub   *redis.PubSub
+}
+
+func NewRedisPubSub(redisUri string) *RedisPubSub {
+	return &RedisPubSub{
+		redisUri: redisUri,
+	}
+}
+
+func (r *RedisPubSub) Start() error {
+	if r.cancel != nil {
+		r.cancel()
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	r.ctx = ctx
+	r.cancel = cancel
+
+	client, err := connectRedis(r.redisUri)
+	if err != nil {
+		return err
+	}
+	r.client = client
+
+	// Note the race between start and stop, not fixing right now
+	pubsub := client.Subscribe(ctx, redisUpsertTopic)
+	r.pubsub = pubsub
+
+	return nil
+}
+
+func (r *RedisPubSub) Stop() error {
+	if r.cancel == nil || r.pubsub == nil || r.client == nil {
+		panic("Stop() called before Start()")
+	}
+
+	r.cancel()
+	r.pubsub.Close()
+	r.client.Close()
+
+	return nil
+}
+
+func (r *RedisPubSub) Subscribe(ctx context.Context) <-chan suave.DAMessage {
+	ch := make(chan suave.DAMessage, 16)
+
+	go func() {
+		for r.ctx.Err() == nil && ctx.Err() == nil {
```

Comment:
if one of these errors is not nil, would it get logged somewhere?

Reply:
generally on this function, i think we might need to add reconnect logic in case the redis connection drops 🤔

Reply:
Will make it an endless loop